Fixing issue #71

dbouianov opened the pull request:
https://github.com/Microsoft/HoloToolkit-Unity/issues/71

The proposed change makes GazeManager.FocusedObject into a public property and introduces several new classes to facilitate sending arbitrary events to dynamic objects via KeywordManager that do not require additional code:

- Adds FocusedObjectMessageSender class whose SendMessageToFocusedObject method can be registered as a response in KeywordManager to enable arbitrary messages to be sent to currently focused object.
- Adds SelectedObjectMessageSender class whose SendMessageToSelectedObject method can be registered as a response in KeywordManager to add voice commands to select objects and enable arbitrary messages to be sent to currently selected object. Saying "Select Object" selects currently focused object and selection persists even when gazing away. Saying "Clear Selection" resets currently  selected object.
- Adds FocusedObjectMessageReceiver class which shows an example on how to handle messages sent by FocusedObjectMessageSender. In this particular example, focused object color is toggled on gaze enter/exit events.
- Adds SelectedObjectMessageReceiver class which shows an example on how to handle messages sent by SelectedObjectMessageSender. In this particular example, selected object color is toggled when selecting the object and clearing currently selected object.
- Adds FocusedObjectKeywords and SelectedObjectKeywords test scenes created to test above functionality. In the first scene, gazing on an object and saying "Make Smaller" and "Make Bigger" will adjust object size. In the second scene, gazing on an object and saying "Select Object" will persistently select that object for interaction with voice commands, after which the user can also adjust object size with "Make Smaller" and "Make Bigger" voice commands and finally clear currently selected object by saying "Clear Selection".

Changes have been discussed in the original PR:
https://github.com/Microsoft/HoloToolkit-Unity/pull/69

This PR also introduces a high-level concept of policies to control object selection:
1) Selecting object for interaction with voice commands by simply gazing at it.
2) Selecting object for interaction with voice commands by gazing at it and issuing an explicit "Select Object" voice command.
3) Other object selection policies are also possible, e.g. auto-aim or selection hysteresis.

msftclas commented:
Hi __@dbouianov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

NeerajW commented on Assets/HoloToolkit/Input/Prefabs/FocusedObjectKeywordManager.prefab:
```diff
@@ -0,0 +1,103 @@
+%YAML 1.1
```
Please don't check in updates to templates.meta

dbouianov replied:
OK, I'll revert that part.

jwittner commented:
@NeerajW made some comments on #69 that pertain here.

NeerajW commented on Assets/HoloToolkit/Input/Prefabs/FocusedObjectKeywordManager.prefab:
```diff
@@ -0,0 +1,103 @@
+%YAML 1.1
+%TAG !u! tag:unity3d.com,2011:
+--- !u!1001 &100100000
```
Could you please update the ReadMe.md on what this prefab is for?
Right now I'm not sure what this does.

dbouianov replied:
Added extensive notes in README.md

NeerajW commented on Assets/HoloToolkit/Input/Tests/Scripts/FocusedObjectMessageReceiver.cs:
```diff
@@ -0,0 +1,31 @@
+/// FocusedObjectMessageReceiver class shows how to handle messages sent by FocusedObjectMessageSender.
+/// This particular implementatoin controls object appearance by changing its color when focused.
+/// </summary>
+public class FocusedObjectMessageReceiver : MonoBehaviour
```
Technically I can achieve the same functionality with just GazeManager.cs.
What are we providing extra with this feature?

jwittner replied:
I think the real value in FocusedObjectMessageSender is in it's place alongside the SelectedObjectMessageSender. And the value of both is that they provide a clear paradigm for dealing with the issue of dynamic objects, gaze, and the immutable nature of the KeywordManager. I really suspect that like @dbouianov others will have the same issue and I do like that the toolkit takes a stance on how to solve this.

Ultimately though this goes back to how we're treating this project. Seems like there's a mix of focus on creating functionality for developers to leverage and focus on instructing or educating developers on how they might do something. This falls into the latter.

NeerajW replied:
I totally support both types of changes. Just want to make sure any ambiguity is resolved wither through comments and tests or not having that code.

I suppose the reason this is advantageous is you can send any custom message to gaze at objects and not just OnGazeEnter and OnGazeLeave.

dbouianov replied:
@jwittner Yes, the value is having the possibility of multiple object selection policies and educating developers about potentially writing their own message senders and receivers, which IMHO keeps the modular spirit of HoloToolkit.

dbouianov replied:
@NeerajW Yes, Having only OnGazeEnter/Leave is limiting. There is another class ScaleObjectMessageReceiver that illustrates how simple message handlers can be added to different objects to enable interesting behaviors in a modular way. For example, to adjust object size, I now simply need to drop FocusedObjectKeywordManager into the scene and add ScaleObjectMessageReceiver onto my objects. I could see a growing library of these simple behaviors added to HoloTookit in the future.

dbouianov replied:
We went back and forth on this in #69, the reason for refactoring this into a separate class is that developers will want different ways of handling object selection, not necessarily based on gaze, so it feels more natural to keep the send message part out of GazeManager, especially that there is also a SelectedObjectMessageSender, which shows a different way to select objects. Selection can also be via airtap, alone or in addition to voice commands.

NeerajW commented on Assets/HoloToolkit/Input/Tests/Scripts/ScaleObjectMessageReceiver.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    public const float DefaultSizeFactor = 2.0f;
+
+    public float SizeFactor = DefaultSizeFactor;
```
Does this need to be public?

dbouianov replied:
Nope

NeerajW commented on ProjectSettings/EditorBuildSettings.asset:
```diff
@@ -4,4 +4,8 @@
   m_ObjectHideFlags: 0
   serializedVersion: 2
-  m_Scenes: []
+  m_Scenes:
```
Please revert this file. 
You are attempting to check in your Editor build settings to the entire project.

dbouianov replied:
OK, reverted.

NeerajW commented on ProjectSettings/GraphicsSettings.asset:
```diff
@@ -3,7 +3,7 @@
 --- !u!30 &1
```
Please revert this file.

dbouianov replied:
OK, done

NeerajW commented on Assets/HoloToolkit/Input/Scripts/SelectedObjectMessageSender.cs:
```diff
@@ -0,0 +1,46 @@
+/// Object selection is controlled via OnSelectObject and OnClearSelection events.
+/// Object selection and messages can be triggered using keyword responses registered in KeywordManager.
+/// </summary>
+public class SelectedObjectMessageSender : MonoBehaviour
```
I guess clarifying here that this selection happens with a voice command Vs an air-tap gesture would be beneficial.
1. Is that true? 
2. Should this work in conjunction with air-tap?
3. If you have air-tap wired up then just saying 'Select' also triggers the same action.

jwittner replied:
Ultimately I like that any number of things can trigger this. Or are you asking how this will be configured in the test scene?

NeerajW replied:
Yeah if multiple can trigger then then showing both examples would be nice.
Or stating that in the class description and readme.

dbouianov replied:
@NeerajW I added more extensive comments about voice commands since the note about KeywordManager responses may not be clear.

jwittner commented on ProjectSettings/ProjectVersion.txt:
```diff
@@ -1,2 +1,2 @@
-m_EditorVersion: 5.4.0b14-HTP
+m_EditorVersion: 5.4.0b18-HTP
```
How do we feel about keeping the project up to date with the Unity drops available? I'm all for anything that pushes this number higher, but it should probably be in a separate pull right?

NeerajW replied:
I like that idea. B20 is the latest.

dbouianov replied:
Looks like B22 is actually the latest now, but this probably needs to be done separately in another PR.

jwittner commented on Assets/HoloToolkit/Input/Tests/Scripts/SimpleGridGenerator.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using System.Collections.Generic;
+using UnityEngine;
+
+public class SimpleGridGenerator : MonoBehaviour
```
Not sure how this fits in here.

NeerajW replied:
I agree. It's a nice idea but don't think we need it really :)

NeerajW replied:
With this file, were you trying to demonstrate that this works with dynamically created prefab instances?

dbouianov replied:
Exactly. If not grid, I had to add some mechanism for placing dynamic objects.

NeerajW replied:
Makes sense. Please do update the readme with such info.

jwittner replied:
I think even if the objects are placed directly in the scene we're still dynamically choosing which one to send the message too based on gaze. I believe this is a clear enough demonstration without needing to actually dynamically add the objects to the scene and removing this would make the pull request a lot simpler.

dbouianov replied:
Issue #71 is not being able to send messages to dynamically created objects, and I was asked to create test scenes to illustrate how this would work in a real scenario. I understand the rationale to keep HoloToolkit as simple as possible, but removing dynamic objects would defeat the purpose of the test case and obscure the intent of the proposal.

I think it would only help developers top have as many examples as possible.

If the desire is to keep HoloToolkit simple, I would suggest adding a separate Examples/Test folder under the root HoloToolkit folder instead of keeping Test folders alongside each module. Then, tests and examples can be excluded when importing a clean version of HoloToolkit into a production project.

jwittner replied:
Ah, I think the real underlying issue isn't whether the objects were created dynamically, but instead the current inability to dynamically decide what object to invoke from a voice command. I believe solving the latter solves the former and demonstrating it with placed objects would be fine.

I believe as you do that we're likely to add more kinds of these helper strategies for command forwarding and think that by keeping the tests to the minimum required for a solution we'll have less to manage going forward. If others find value in the dynamic allocation I'm OK with keeping it, I just don't think it's necessary.

NeerajW replied:
Just a comment to clarify this:
>> If the desire is to keep HoloToolkit simple, I would suggest adding a separate Examples/Test folder under the root HoloToolkit folder instead of keeping Test folders alongside each module.

We started like this but then we had cross folder dependencies which was not fun. This way developers can completely delete the folder feature areas that they don't rely on.

dbouianov replied:
Could you share what wasn't fun about keeping cross-folder dependencies in separate folders?

This is what I thought:
HoloToolkit/UnitTests/Input - would depend only on HoloToolkit/Input
HoloToolkit/IntegrationTests/Input - could have cross-folder dependencies
HoloToolkit/Examples/Input - could have cross-folder dependencies

I think additional examples and tests could probably outweigh a potential inconvenience of having to delete a few more folders once in a project lifecycle.

Another way to do this is potentially create a separate repository (say HoloToolkit-Unity-Examples) which could point to HoloToolkit-Unity as a submodule.

NeerajW commented:
Curious, what do you think of our general over reliance on the Unity SendMessage paradigm? Performance impacts?

jwittner commented:
@NeerajW Personally I don't like the message paradigm because it defines how you interact with something in the implementation instead of at an API level. You can't look at the API of a class and know how to work with it, you have to rely on documentation and I think that's sad. 

Regarding performance I'm unsure on the underlying implementation, but I suspect it's slow and relies on reflection. It'd be interesting to run some tests on SendMessage vs a virtual method call on an interface.

dbouianov commented:
@NeerajW @jwittner I've had to create strongly-typed message passing mechanisms for other Unity projects in the past. There is a lot of online debate on the topic of message passing in Unity, and it might be a challenge to introduce such a mechanism into HoloToolkit without making into an opinionated framework. String messages are easy, but not performant and prone to error, but callbacks need a lot of boilerplate when dealing with dynamic objects. Factory classes that register keyword handler callbacks for families of dynamic objects would probably introduce too much boilerplate, while arbitrary messages are nice because they make for very loose coupling between issued keywords and objects handling those keywords.

dbouianov commented:
OK, addressed code review comments, added a lot more documentation and comments, removed some unnecessary files and reverted some files to earlier versions.

NeerajW commented:
>>while arbitrary messages are nice because they make for very loose coupling between issued keywords and objects handling those keywords.

This was one of the main reasons why we went with this approach.

jwittner commented:
@NeerajW Are we squashing changes when we merge PRs? If not I think we should recommend a rebase on the history for this change.

dbouianov commented:
OK, I can rebase again.

NeerajW commented:
@dbouianov thank you for all the updates. Look forward to more of them. Merged.
